syntax = "proto3";
package plugin;

option go_package = "github.com/gideaworx/terraform-exporter-plugin/go-plugin/proto";
option java_multiple_files = true;
option java_package = "io.gideaworx.terraformexporterplugin.grpc";

message ExportRequest {
  string OutputDirectory = 1;
  bool SkipProviderOutput = 2;
  repeated string PluginArgs = 3;
}

message ImportDirective {
  string Resource = 1;
  string Name = 2;
  string ID = 3;
}

message Version {
  uint64 Major = 1;
  uint64 Minor = 2;
  uint64 Patch = 3;
  optional string Pre = 4;
  optional string BuildMetadata = 5;
}

message CommandInfo {
  string Name = 1;
  string Description = 2;
  string Summary = 3;
  Version Version = 4;
}

message PluginInfo {
  Version Version = 1;
  repeated CommandInfo Provides = 2;
}

message ExportResponse {
  repeated ImportDirective Directives = 1;
}

message Empty {}

message SingleString {
  string Value = 1;
}

message PluginRequest {
  string Name = 1;
  ExportRequest Request = 2;
}

service ExportCommand {
  rpc Export(ExportRequest) returns (ExportResponse) {}
  rpc Info(Empty) returns (CommandInfo) {}
  rpc Help(Empty) returns (SingleString) {}
}

service ExportPlugin {
  rpc Export(PluginRequest) returns (ExportResponse) {}
  rpc Help(SingleString) returns (SingleString) {}
  rpc Info(Empty) returns (PluginInfo) {}
}
